synapse/api/room_versions.py

```diff
@@ -58,7 +58,10 @@ class PushRuleRoomFlag:
     EXTENSIBLE_EVENTS = "org.matrix.msc3932.extensible_events"
 
 
-@attr.s(slots=True, frozen=True, auto_attribs=True)
+# eq=False means equality and hashing are done using object identity, which suits
+# our purposes because each RoomVersion is instantiated only once and some of
+# the fields are not hashable.
+@attr.s(slots=True, frozen=True, auto_attribs=True, eq=False)
```

Comment:
Could also consider changing `msc3931_push_features` to `Tuple[str, ...]` instead---seems less scary to me, but maybe you disagree.

Reply:
Basically I'm scared that `RoomVersion is instantiated only once` might become False in the future

Reply:
That's a fair point. Nobody should be mutating that list anyway.

Reply:
> Basically I'm scared that `RoomVersion is instantiated only once` might become False in the future

(I would say it doesn't seem likely given all the room versions are introduced as constants, but you're right that we don't know what the future holds..)

Reply:
now done